### ISSUE TITLE
`tail`: Intermittent failure of `test_tail::test_follow_name_truncate4` on freebsd

### ISSUE THREAD

**Issue by Joining7943** [uutils/coreutils#4267]
Happened here https://github.com/uutils/coreutils/actions/runs/3862874468/jobs/6584760432

```
  ---- test_tail::test_follow_name_truncate4 stdout ----
  current_directory_resolved: 
  write(append): /tmp/.tmpxN77vP/file
  run: /Users/runner/work/coreutils/coreutils/target/debug/coreutils tail -s.1 --max-unchanged-stats=1 -F file
  write(truncate): /tmp/.tmpxN77vP/file
  thread 'test_tail::test_follow_name_truncate4' panicked at 'Expected stderr to be empty, but it's:
  tail: file: file truncated
  ', tests/by-util/test_tail.rs:2026:14
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/std/src/panicking.rs:575:5
     1: core::panicking::panic_fmt
               at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/panicking.rs:65:14
     2: tests::common::util::CmdResult::no_stderr
               at ./tests/common/util.rs:414:9
     3: tests::common::util::CmdResult::stdout_only
               at ./tests/common/util.rs:585:9
     4: tests::test_tail::test_follow_name_truncate4
               at ./tests/by-util/test_tail.rs:2023:9
     5: tests::test_tail::test_follow_name_truncate4::{{closure}}
               at ./tests/by-util/test_tail.rs:2003:1
     6: core::ops::function::FnOnce::call_once
               at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
     7: core::ops::function::FnOnce::call_once
               at /rustc/69f9c33d71c871fc16ac445211281c6e7a340943/library/core/src/ops/function.rs:251:5
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by cre4ture** [comment]
happend again:
https://github.com/uutils/coreutils/actions/runs/8529289915/job/23364853543?pr=6154

```
---- test_tail::test_follow_name_truncate4 stdout ----
write(append): /tmp/.tmpJ7XcP6/file
run: /target/i686-unknown-linux-musl/debug/coreutils tail -s.1 --max-unchanged-stats=1 -F file
write(truncate): /tmp/.tmpJ7XcP6/file
thread 'test_tail::test_follow_name_truncate4' panicked at 'Expected stderr to be empty, but it's:
tail: file: file truncated
', tests/by-util/test_tail.rs:2006:14
stack backtrace:
```

**Comment by dan-hipschman** [comment]
I've noticed this on a few recently PRs as well, although they were all in the SELinux build. Here are some examples:
* https://github.com/uutils/coreutils/actions/runs/14339428856/job/40195456309?pr=7686#step:11:4107
* https://github.com/uutils/coreutils/actions/runs/14351701708/job/40231924636#step:11:4111

These tail tests start with a file, then truncates it with the same contents, and expects no event to be triggered:
https://github.com/uutils/coreutils/blob/9cb4348a5583f0128bf02f7fa17e391945df4804/tests/by-util/test_tail.rs#L1980-L1993

I think the problem is that truncating and writing isn't atomic. It first truncates the file, then writes new contents to it:
https://github.com/uutils/coreutils/blob/9cb4348a5583f0128bf02f7fa17e391945df4804/tests/uutests/src/lib/util.rs#L1006-L1016

So there's a window where the test can fail, and I'm guessing it happens more on platforms that are tested within VMs (such as SELinux) because it's slower or inotify events are just implemented differently within the VM.

As a potential solution, I think if `tail` is being tested in `--follow=name` mode, the test could replace `AtPath.truncate(file, contents)` with a new `AtPath.replace(file, contents)` that writes contents to a new temp file and then renames it to the specified file. I haven't tried but I believe this should fix it. If this sounds reasonable then I could make a PR.

An alternative would be to increase the sleep interval passed to `tail`. It's currently 100ms which seems more than enough, but maybe not. And I'm not sure if this is even used with inotify either, is it only used when polling?

**Comment by dan-hipschman** [comment]
I looked into this some more, and have come to the conclusion that tail (both GNU and uutils) is just inherently nondeterministic when using inotify. Hence, ignore my comment above.

Here's an example. Assume file `hello` contains the string `world`, and we run `tail -f hello`. As expected, it outputs `world` and then waits for more lines in the file. Then assume we run `echo world > hello` in another terminal. Now two things could happen: a) there's no new output, or b) tail writes that the file is truncated to stderr and outputs a second `world`.

In case (a) tail receives the inotify event after `world` has been written to the file. It notices the event, but there's no new output since the file size is the same. In case (b) tail receives an inotify event after the file is truncated but before `world` is written. Since the file size decreased, it assumes the file was truncated and reopens it. Hence it outputs `world` a second time. Note that tail will always receive multiple inotify events, it just depends *when* they are received.

It actually doesn't matter what the contents are, as tail only checks the file size. Hence, writing different contents to the file will likely be ignored unless the inotify event gets seen between the truncate and write. Here's an example:

In one shell run `echo > /tmp/test ; tail -f /tmp/test`. In another run `for i in (seq 10) ; printf "%02d\n" $i > /tmp/test ; end` (this is fish syntax but hopefully is clear). Here's what I see (GNU and uutils are both similar):
```
$ tail -f /tmp/test

1
tail: /tmp/test: file truncated
05
tail: /tmp/test: file truncated
07
tail: /tmp/test: file truncated
09
tail: /tmp/test: file truncated
10
```
Note the first `0` from `01` is cut off because tail has already output 1 newline byte so ignores the first byte of the updated file.

This actually also "works" when the file size increases. For instance, setup the same test but this time run `for s in a bb ccc dddd eeeee ffffff ; echo $s > /tmp/test ; end`. Here's the output I see:
```
$ tail -f /tmp/test
a
tail: /tmp/test: file truncated
dddd
```
I'm not actually sure what exactly is happening here, but the point is this is nondeterministic.

So basically, there's just some race condition which makes it nondeterministic whether tail notices the file is truncated or not. This makes deterministic tests hard to write. Maybe the tests can gracefully accept multiple outputs, or they could be rerun on failure to reduce the likelihood of "unlucky" timing causing CI failures. Or someone could fix GNU tail to be deterministic and then follow suit in uutils. :)

**Comment by sylvestre** [comment]
Verified fixed - test_tail::test_follow_name_truncate4 now passes (commit f551cc458)
